### PR TITLE
Away notify

### PR DESCRIPTION
**Pull request by ThePhilgrim**
This PR includes functionality regarding the `AWAY` command as well as the capability `away-notify`.

AWAY:
- When a user sends `/AWAY <Away message>`, they will receive a message in each channel view, saying that they have been marked as away
- When a user sends `/AWAY`, they will receive a message in each channel view, saying that they are no longer marked as being away

AWAY-NOTIFY:
- When `user Foo` sends an `/AWAY` command, `user Bar` will be notified about this, if they have negotiated the `away-notify` capability with the server during registration.
- A user who has negotiated the `away-notify` capability with the server during registration will automatically send a `/WHO` command when joining a new channel. The response from the server looks something like this:

:tantalum.libera.chat 352 fubbe ##learnpython ~thephilgr 141.37.165.190 mercury.libera.chat ThePhilgrim `H / G` :0 ThePhilgrim
`H` = Not away
`G`= Away

All users with their status set to `AWAY` will have their nick greyed out in channel views.

Implementing this logic required me to modify the capability negotiation a bit. I believe this logic for capability negotiation is also more scalable. There is still some work to be done on this, but it would need to be in a separate PR.

### Left to do:
1. All/most tests are currently failing. After updating tests/Mantatail, I get a RuntimeError that I cannot seem to figure out the reason for:

```
# Wait for it to start
for line in self.process.stdout:
        assert b"error" not in line.lower()
        if b"Starting hircd" in line or b"Mantatail running" in line:
                break
        else:
                raise RuntimeError <-----
```

After this has been solved, I can continue the implementation of tests relevant to this PR.

2. Several changes have been made on `master` branch. This PR currently has merge conflicts.

3. Update README to say Mantaray supports `AWAY`

### Potential issues:
When autojoining many channels upon connection, several `WHO` commands will be sent. This can make servers such as Libera close the connection due to "excessive flooding". I seem to have manage to solve it with the way these `WHO` commands are sent now, but we should keep an eye on it. According to the guys over at #ircv3, the best way to avoid this issue is to wait for the `WHOREPLY` before sending the next `WHO`. I am not yet completely sure how to code this logic, but if needed I will try to figure it out.

### CODE REVIEW

**Comment by ThePhilgrim** [comment]
TL;DR: away-notify, yay!

File: mantaray/received.py [hunk 1]
```diff
@@ -228,6 +234,20 @@ def _handle_quit(server_view: views.ServerView, nick: str, args: list[str]) -> N
             view.userlist.remove_user(nick)
 
 
+def _handle_away(server_view: views.ServerView, nick: str, args: list[str]) -> None:
+    for view in _get_views_relevant_for_nick(server_view, nick):
+        if not args:
+            view.add_message(
+                "*", (nick, ["other-nick"]), (" is no longer away.", ["info"])
```
Comment (Akuli) [inline mantaray/received.py:241]: I no longer think this is a good idea. It will result in a lot of spam in a big channel, and doesn't solve an actual problem in small channels. But we should update the userlist colors.

Also, I recently changed how `add_message()` should be called, so this needs some updating :)

File: mantaray/received.py [hunk 2]
```diff
@@ -312,12 +332,27 @@ def _handle_cap(server_view: views.ServerView, args: list[str]) -> None:
...
+
+    server_view.core.cap_req.pop()  # To evaluate how many more ACK/NAKs will be received from server
+
+    # Need to wait for login success/fail from server if sasl is enabled before sending "CAP END"
+    if len(server_view.core.cap_req) == 0 and "sasl" not in server_view.core.cap_list:
+        _send_cap_end(server_view)
+
+
+def _send_cap_end(server_view: views.ServerView) -> None:
+    server_view.core.send("CAP END")
```
Comment (Akuli) [inline mantaray/received.py:355]: I don't like short functions like this. Feel free to write them into your own projects, but in my projects, if there's a one-line send, I want to just have that one-line send, not hidden inside a function.

Also applies to `send_who()` method in `backend.py`.

File: mantaray/config.py [hunk 3]
```diff
@@ -217,6 +219,20 @@ def __init__(
             )
             self._rownumber += 1
 
+        self._away_notify_var = tkinter.BooleanVar(value=initial_config["away_notify"])
+        if connecting_to_new_server:
+            self.away_notify_checkbox = None
+        else:
+            self.away_notify_checkbox = ttk.Checkbutton(
+                self,
+                text="Notify me about other users' Away status",
```
Comment (Akuli) [inline mantaray/config.py:228]: I think we can just always request the `"away-notify"` capability. If the server doesn't have it, we won't be able to update the user list colors. I don't see any reason why you would want to disable that, so we no longer need a checkbox in settings.

File: mantaray/commands.py [hunk 4]
```diff
@@ -154,6 +156,24 @@ def kick(view: View, core: IrcCore, nick: str, reason: str | None = None) -> Non
         else:
             view.add_message("You can use /kick only on a channel.", tag="error")
 
+    def away(view: View, core: IrcCore, away_message: str | None = None) -> None:
+        if away_message is None:
+            core.send("AWAY")
+            for view in view.server_view.get_subviews():
+                view.add_message(
+                    "*", ("You are no longer marked as being away", ["info"])
```
Comment (Akuli) [inline mantaray/commands.py:164]: I would prefer a separate command `/back` to be used for this. It's not very intuitive that typing `/away` makes you not be away. It's the complete opposite of what you'd expect it to do.

File: mantaray/backend.py [hunk 5]
```diff
@@ -115,6 +115,11 @@ def __init__(self, server_config: config.ServerConfig, *, verbose: bool):
         self.event_queue: queue.Queue[IrcEvent] = queue.Queue()
         self._thread: threading.Thread | None = None
 
+        # Will contain the capabilities to negotiate with the server
+        self.cap_req: list[str] = []
+        # "CAP LIST" shows capabilities enabled on the client's connection
+        self.cap_list: Set[str] = set()
```
Comment (Akuli) [inline mantaray/backend.py:121]: These lists should probably be cleared when connecting to the server. Otherwise they can contain whatever was left over from the previous connection, if reconnecting happens in the middle of capability negotiation, for example.

File: mantaray/received.py [hunk 6]
```diff
@@ -312,12 +332,27 @@ def _handle_cap(server_view: views.ServerView, args: list[str]) -> None:
...
+            server_view.core.cap_list.add("away-notify")
     elif subcommand == "NAK":
         rejected = set(args[-1].split())
         if "sasl" in rejected:
             # TODO: this good?
             raise ValueError("The server does not support SASL.")
+        if "away-notify" in rejected:
+            raise ValueError("The server does not support away-notify.")
+
+    server_view.core.cap_req.pop()  # To evaluate how many more ACK/NAKs will be received from server
```
Comment (Akuli) [inline mantaray/received.py:347]: I don't like how responding to a capability `foo` can cause capability `bar` to be popped from the list. Can you replace `cap_req` with an integer that counts the number of accepted/rejected capabilities?

File: mantaray/received.py [hunk 7]
```diff
@@ -312,12 +332,27 @@ def _handle_cap(server_view: views.ServerView, args: list[str]) -> None:
...
             server_view.core.send("AUTHENTICATE PLAIN")
+        if "away-notify" in acknowledged:
+            server_view.core.cap_list.add("away-notify")
     elif subcommand == "NAK":
         rejected = set(args[-1].split())
         if "sasl" in rejected:
             # TODO: this good?
             raise ValueError("The server does not support SASL.")
+        if "away-notify" in rejected:
+            raise ValueError("The server does not support away-notify.")
```
Comment (Akuli) [inline mantaray/received.py:345]: Mantaray should be able to work with servers that don't support away-notify. It shouldn't be an error. It just means you won't get colors in the userlist, and that's all it does.

**Review by Akuli — COMMENTED** [review]
I have thought about this a bit more, and realized that I want mantaray to behave differently than I initially said. This means you will have to delete and edit some stuff. Sorry about that!

File: mantaray/commands.py [hunk 8]
```diff
@@ -154,6 +156,30 @@ def kick(view: View, core: IrcCore, nick: str, reason: str | None = None) -> Non
...
+    def away(view: View, core: IrcCore, away_message: str | None = None) -> None:
+        if away_message is None:
+            current_view = view.server_view.irc_widget.get_current_view()
+            current_view.add_message(
+                "*",
+                (
+                    'Please use "/AWAY <away message>" to mark yourself as Away, or "/BACK" to remove your Away status.',
+                    ["error"],
+                ),
+            )
```
Comment (Akuli) [inline mantaray/commands.py:168]: You don't need this code, if you define `away_message: str` without a default value. Then `/away` with no message will produce an error.

File: mantaray/received.py [hunk 9]
```diff
@@ -312,13 +327,27 @@ def _handle_cap(server_view: views.ServerView, args: list[str]) -> None:
     if subcommand == "ACK":
         acknowledged = set(args[-1].split())
         if "sasl" in acknowledged:
+            server_view.core.cap_list.add("sasl")
             server_view.core.send("AUTHENTICATE PLAIN")
+        if "away-notify" in acknowledged:
+            server_view.core.cap_list.add("away-notify")
```
Comment (Akuli) [inline mantaray/received.py:333]: Can we just add all acknowledged capabilities, instead of checking each one manually?

File: mantaray/received.py [hunk 10]
```diff
@@ -312,13 +327,27 @@ def _handle_cap(server_view: views.ServerView, args: list[str]) -> None:
...
     elif subcommand == "NAK":
         rejected = set(args[-1].split())
         if "sasl" in rejected:
             # TODO: this good?
             raise ValueError("The server does not support SASL.")
 
+    server_view.core.number_of_cap_req -= (
+        1  # To evaluate how many more ACK/NAKs will be received from server
```
Comment (Akuli) [inline mantaray/received.py:341]: This comment IMO belongs to where `number_of_cap_req` is defined. I also don't like the name. How about `pending_capabilities` or `pending_cap_count`, pending as in we don't know yet whether server has it or not?

File: mantaray/received.py [hunk 11]
```diff
@@ -312,13 +327,27 @@ def _handle_cap(server_view: views.ServerView, args: list[str]) -> None:
...
+    )
+
+    # Need to wait for login success/fail from server if sasl is enabled before sending "CAP END"
```
Comment (Akuli) [inline mantaray/received.py:344]: I don't understand this comment. What exactly is "enabling" SASL? Why do we need to wait? What happens if we don't wait?

File: mantaray/views.py [hunk 12]
```diff
@@ -20,6 +20,7 @@
 class _UserList:
     def __init__(self, irc_widget: IrcWidget):
         self.treeview = ttk.Treeview(irc_widget, show="tree", selectmode="extended")
+        self.treeview.tag_configure("away", foreground="#95968c")
```
Comment (Akuli) [inline mantaray/views.py:23]: Nick changes currently mark the user as not away, because it removes the user and adds again but with no tags. Can fix in separate PR.

File: mantaray/received.py [hunk 13]
```diff
@@ -228,6 +235,14 @@ def _handle_quit(server_view: views.ServerView, nick: str, args: list[str]) -> N
             view.userlist.remove_user(nick)
 
 
+def _handle_away(server_view: views.ServerView, nick: str, args: list[str]) -> None:
+    for view in _get_views_relevant_for_nick(server_view, nick):
+        if not args:
+            view.userlist.treeview.item(nick, tag=[])
+        else:
+            view.userlist.treeview.item(nick, tag=["away"])
```
Comment (Akuli) [inline mantaray/received.py:243]: Can you add a new method to the userlist,  `set_away(nick: str, away: bool)` or something like that? Code outside `class _UserList` shouldn't need to access the treeview directly.

File: mantaray/received.py [hunk 14]
```diff
@@ -115,11 +120,13 @@ def _handle_privmsg(
 
 
 def _handle_join(server_view: views.ServerView, nick: str, args: list[str]) -> None:
+    [channel] = args
     # When this user joins a channel, wait for RPL_ENDOFNAMES
     if nick == server_view.core.nick:
+        if "away-notify" in server_view.core.cap_list:
+            server_view.core.send(f"WHO {channel}")
```
Comment (Akuli) [inline mantaray/received.py:127]:
```suggestion
            # TODO: sends WHOs too frequenty, should wait for WHOREPLY
            server_view.core.send(f"WHO {channel}")
```

**Comment by Akuli** [comment]
I fixed the Git mess. The lesson is, don't revert a merge, because it adds back a pile of changes that are difficult to get rid of. A better approach might be `git reset --hard good-commit-id` followed by `git push --force`. You have to be careful with this, because it's possible to lose commits forever with this command. This is why I usually prefer reverting. Both approaches are explained more detail in stackoverflow (google e.g. how to revert a commit).

What still needs to be done:

1. Make mantaray work with IRC servers that don't know anything about the `CAP` command. These servers don't say "I don't have this capability", they say "unknown command CAP". One of them is `hircd`. It comes with mantaray as a git submodule, just like mantatail, and there's instructions for running tests with hircd in README. If you want to start hircd manually, use this command: https://github.com/Akuli/mantaray/blob/d2933008d326c1ec4ad2ad701af38234e9a2aed8/tests/conftest.py#L63-L69 (`sys.executable` basically means `python3`)
2. Get GitHub Actions green. It runs tests with both hircd and mantatail.
3. Add tests for the new functionality. It is possible to check what tags a given item has in the user list. Mantatail should support everything we need, but hircd does not, so you will need to use `pytest.mark.skipif`. There's already plenty of other tests that are skipped on hircd.

File: mantaray/commands.py (hunk 8 above)
Comment (ThePhilgrim) [reply mantaray/commands.py:168]: If I understand it correct, we still want to add this as a message to the user though so they know why their command is not working.

Therefore I'd say that the code I don't need is `if away_message is None`. I will instead use a `try/except`

Comment (ThePhilgrim) [reply mantaray/commands.py:168]: @Akuli Since your fixes `/away` is broken.
`/AWAY` gives message `Usage: /AWAY <away_message>`, but it doesn't have an error tag (which i think it should, to stand out in a busy channel).

`/AWAY Foo` doesn't clear the entry and you don't get the message that you are marked as being away. Same with `/BACK`.

Comment (ThePhilgrim) [reply mantaray/commands.py]: I understand now that you implemented a "Usage" error thingy. I would say this should have the error tag instead of info though. Not sure if all "Usage" messages should be errors though, so I'll leave this up to you.

File: mantaray/backend.py [hunk 15]
```diff
@@ -200,7 +205,16 @@ def _connect_loop(self) -> None:
                 continue
 
             if self.password is not None:
-                self.send("CAP REQ sasl")
+                self.cap_req.append("sasl")
+
+            self.cap_req.append("away-notify")
+
+            self.pending_cap_count = len(
```
Comment (Akuli) [inline mantaray/backend.py:212]: This should be set to 0 in `__init__`. It's where people expect attributes to be defined. And yes, I know, there's `_apply_config()`, but that method is just below `__init__` for a reason.

File: mantaray/backend.py [hunk 16]
```diff
@@ -200,7 +205,16 @@ def _connect_loop(self) -> None:
                 continue
 
             if self.password is not None:
-                self.send("CAP REQ sasl")
+                self.cap_req.append("sasl")
```
Comment (Akuli) [inline mantaray/backend.py:208]: Presumably `cap_req` and `cap_list` should be cleared somewhere in the middle of `_connect_loop`, before you append more capabilities. It is currently done when receiving `RPL_WELCOME`, but there's no way that code in different files will stay in sync when we add a new list that also needs to be cleared.

File: mantaray/commands.py [hunk 17]
```diff
@@ -154,6 +156,20 @@ def kick(view: View, core: IrcCore, nick: str, reason: str | None = None) -> Non
         else:
             view.add_message("You can use /kick only on a channel.", tag="error")
 
+    def away(view: View, core: IrcCore, away_message: str) -> None:
+        core.send(f"AWAY :{away_message}")
+        for view in view.server_view.get_subviews(include_server=True):
+            view.add_message("You have been marked as being away", tag="info")
```
Comment (Akuli) [inline mantaray/commands.py:162]: `tag="info"` is unnecessary, because the `tag` argument has `"info"` as its default value.

File: mantaray/received.py [hunk 18]
```diff
@@ -463,13 +507,21 @@ def _handle_received_message(
...
     elif msg.command == "AUTHENTICATE":
         _handle_authenticate(server_view)
 
-    elif msg.command == RPL_LOGGEDIN:
+    elif msg.command == RPL_WELCOME:
+        server_view.core.cap_req = []
+        server_view.core.cap_list = set()
+
+    elif msg.command == RPL_SASLSUCCESS or msg.command == ERR_SASLFAIL:
         server_view.core.send("CAP END")
```
Comment (Akuli) [inline mantaray/received.py:526]: We should probably add the received text into the server view. Currently this doesn't display an error message anywhere when it receives `ERR_SASLFAIL`.

**Review by Akuli — COMMENTED** [review]
A couple new thoughts from skimming through the code:

File: mantaray/commands.py (hunk 8 above)
Comment (Akuli) [reply mantaray/commands.py]: I created #239.

File: mantaray/commands.py [hunk 19]
```diff
@@ -89,6 +89,8 @@ def _define_commands() -> dict[str, Callable[..., None]]:
     def join(view: View, core: IrcCore, channel: str) -> None:
         # TODO: plain '/join' for joining the current channel after kick?
         core.send(f"JOIN {channel}")
+        if "away-notify" in core.cap_list:
+            core.send(f"WHO {channel}")
```
Comment (Akuli) [inline mantaray/commands.py:93]: This will send `WHO #css`, if you send `JOIN #css` but the server redirects you to `#css-overflow`. We should probably send `WHO` in `_handle_endofnames()`.

File: mantaray/commands.py [hunk 20]
```diff
@@ -154,6 +156,20 @@ def kick(view: View, core: IrcCore, nick: str, reason: str | None = None) -> Non
         else:
             view.add_message("You can use /kick only on a channel.", tag="error")
 
+    def away(view: View, core: IrcCore, away_message: str) -> None:
+        core.send(f"AWAY :{away_message}")
+        for view in view.server_view.get_subviews(include_server=True):
+            view.add_message("You have been marked as being away")
```
Comment (Akuli) [inline mantaray/commands.py:162]: The IRC server responds to `AWAY :message` with 306, and to just `AWAY` with 305. We should display those to the user instead of our own message. This way, if the server doesn't support `AWAY`, the user won't get a confusing message saying that their `/away` succeeded.

File: mantaray/received.py [hunk 21]
```diff
@@ -316,14 +332,31 @@ def _handle_cap(server_view: views.ServerView, args: list[str]) -> None:
...
+        for capability in acknowledged:
+            server_view.core.cap_list.add(capability)
+
     elif subcommand == "NAK":
         rejected = set(args[-1].split())
         if "sasl" in rejected:
             # TODO: this good?
             raise ValueError("The server does not support SASL.")
 
+    server_view.core.pending_cap_count -= 1
```
Comment (Akuli) [inline mantaray/received.py:348]:
```suggestion
    server_view.core.pending_cap_count -= len(rejected)
```

Or if you are sure that it cannot NAK multiple capabilities at once, change the code above instead.

**Review by Akuli — COMMENTED** [review]
The final nits...

**Comment by ThePhilgrim** [comment]
> 1. Make mantaray work with IRC servers that don't know anything about the `CAP` command.

@Akuli Mantaray already works with servers that don't know anything about the `CAP` command. There is no code in Mantaray that is dependent on the server to respond to the `CAP REQ`. 

It will continue to connect without negotiating capabilities, without any issues.

File: mantaray/received.py (hunk 21 above)
Comment (ThePhilgrim) [reply mantaray/received.py:348]: @Akuli It can NAK multiple capabilities at once if you do `CAP REQ capability_1 capability_2 capability_3`.

Also, if we get an `ACK`, `pending_cap_count` gets `-= 0`, then it will never send `CAP END`. It has to do at least `-= 1`

**Comment by ThePhilgrim** [comment]
@Akuli I can't seem to understand why there is a `RuntimeError` when running the hircd tests. Hircd should just give a simple `421 Unknown command` and then everything should proceed as usual.
In fact, if I connect to hircd with `mantaray --alice` and I add a password, Hircd still sends the 421. This means that the issue is not really introduced in this PR, it just reveals an already existing bug.

At this point, I would also consider not running the tests with hircd anymore. Mantatail supports everything that hircd supports and more. It's a little bit as if you would run tests with Libera, and still run tests with Mantatail just for the sake of doing it.

If you still want to run tests with hircd, would you mind committing a fix for the `RuntimeError` that it produces?

**Comment by Akuli** [comment]
I fixed the hircd tests.

> At this point, I would also consider not running the tests with hircd anymore.

I understand that you say this, because you wrote mantatail :)

> Mantatail supports everything that hircd supports and more.

This is true, but not a good reason to remove hircd. I see this as a reason to **keep** testing with hircd. Mantaray should work with all kinds of IRC servers, even IRC servers that don't implement much. For example, it shouldn't refuse to connect if the server doesn't know about the `CAP` command.

> This means that the issue is not really introduced in this PR, it just reveals an already existing bug.

I don't understand how you arrived at this conclusion. The test raises `RuntimeError` when the hircd output mentions "error", which it now started doing because this PR makes hircd send `CAP: unknown command` errors. I made the test to not complain about that error, as I have already done with several other errors.

File: mantaray/received.py (hunk 21 above)
Comment (Akuli) [reply mantaray/received.py:351]: So it should treat `args[-1]` as a single capability, instead of splitting it?

Comment (Akuli) [reply mantaray/received.py:351]: Another thing I just noticed: The `if`,`elif` has no `else`, but it should probably have one. If the server sends `CAP FOO`, it should be either ignored silently, or it should be an error. Currently that would decrement `pending_cap_count` and potentially cause weird problems.

**Comment by ThePhilgrim** [comment]
> I understand that you say this, because you wrote mantatail :)

This isn't relevant though. I write mantatail for learning and improving. It doesn't matter to me if you use mantatail or hircd or something else. My point was that if Hircd supports features "foo", and mantatail supports features "foo" + "bar". Then hircd doesn't really fill a function. However, as you say, it makes sense to do tests on a server with less features as well to make sure Mantaray doesn't ever break. 

> I don't understand how you arrived at this conclusion.

I'm trying to say that the only reason the tests passed were because Alice & Bob don't have passwords in their configs. But the bug was still there because if you were to develop Mantaray connecting to hircd and you had a password entered, you would run into the same problem. In that way, the issues were previously only avoided, not solved.

File: tests/test_commands.py [hunk 22]
```diff
@@ -234,6 +233,62 @@ def test_nickserv_and_memoserv(alice, bob, wait_until):
...
+
+    switch_to_channel_view(alice, "#autojoin")
+
+    # Channel view
+    wait_until(lambda: "You have been marked as being away\n" in alice.text())
+
+    assert "away" in alice.get_current_view().userlist.treeview.item("Alice")["tags"]
+
+    # TODO: Get this to work
+    # assert "away" in bob.get_current_view().userlist.treeview.item("Alice"))
```
Comment (Akuli) [inline tests/test_commands.py:257]: The problem with this assert is that commands aren't immediate. The commands go through a send queue and a couple different socket connections, and all that takes time.

But what matters is that Alice will eventually show up as being away. That's what `wait_until()` does:

```suggestion
    wait_until(lambda: "away" in bob.get_current_view().userlist.treeview.item("Alice")["tags"])
```

The `wait_until()` call will error if `"away"` doesn't appear in Alice's tags within 5 seconds. The test passes on my computer with this change applied.

File: tests/test_commands.py [hunk 23]
```diff
@@ -234,6 +233,62 @@ def test_nickserv_and_memoserv(alice, bob, wait_until):
...
+    reason="hircd doesn't support away notifications",
+)
+def test_away_status(alice, bob, wait_until):
+    switch_to_channel_view(bob, "#autojoin")
+
+    alice.entry.insert("end", "/away foo bar baz")
+    alice.on_enter_pressed()
+
+    # Server view
+    wait_until(lambda: "You have been marked as being away\n" in alice.text())
```
Comment (Akuli) [inline tests/test_commands.py:247]: This in fact checks the channel view. If you want to check the server view, you can use `alice.get_server_views()[0].textwidget.get("1.0", "end")` instead of `alice.text()`. Or add a `.text()` method to `class View`, so that you can do simply `alice.get_server_views()[0].text()`.

File: tests/test_commands.py [hunk 24]
```diff
@@ -234,6 +233,62 @@ def test_nickserv_and_memoserv(alice, bob, wait_until):
     wait_until(lambda: "send Bob hello there\n" in bob.text())
 
 
+@pytest.mark.skipif(
+    os.environ["IRC_SERVER"] == "hircd",
+    reason="hircd doesn't support away notifications",
+)
+def test_away_status(alice, bob, wait_until):
+    switch_to_channel_view(bob, "#autojoin")
```
Comment (Akuli) [inline tests/test_commands.py:241]: This does nothing. The test is showing the `#autojoin` channel by default. (You can check this by inserting `import pdb; pdb.set_trace()` to the beginning of the test). There's another unnecessary `switch_to_channel_view()` call below.

**Review by Akuli — COMMENTED** [review]
Here are a few comments that hopefully clarify how the tests work :)

I am on libera now, and I will come there earlier tomorrow.

**Comment by ThePhilgrim** [comment]
@Akuli I think this might be ready for the final review.

File: mantaray/received.py [hunk 25]
```diff
@@ -316,14 +335,35 @@ def _handle_cap(server_view: views.ServerView, args: list[str]) -> None:
...
         acknowledged = set(args[-1].split())
+
         if "sasl" in acknowledged:
             server_view.core.send("AUTHENTICATE PLAIN")
+
+        for capability in acknowledged:
+            server_view.core.cap_list.add(capability)
+
     elif subcommand == "NAK":
         rejected = set(args[-1].split())
```
Comment (Akuli) [inline mantaray/received.py:346]: If I understood correctly, this splitting doesn't make sense as the server NAKs or ACKs only a single capability at a time.

File: mantaray/received.py [hunk 26]
```diff
@@ -386,6 +431,19 @@ def _handle_endofmotd(server_view: views.ServerView) -> None:
...
+def _handle_whoreply(
+    server_view: views.ServerView, command: str, args: list[str]
+) -> None:
+    assert len(args) == 8
+    nick = args[5]
+    away_status = args[6][0]
+    view = server_view.find_channel(args[1])
+
+    assert view is not None
+    if away_status.lower() == "g":
```
Comment (Akuli) [inline mantaray/received.py:443]: I would assert that it's either `"g"` or `"h"`. If it's something else, it will currently assume the user isn't away, which could be confusing to debug.

File: mantaray/received.py [hunk 27]
```diff
@@ -316,14 +335,35 @@ def _handle_cap(server_view: views.ServerView, args: list[str]) -> None:
...
+    else:
+        server_view.core.send("CAP END")
+        raise ValueError("Invalid CAP response. Aborting Capability Negotiation.")
+
+    server_view.core.pending_cap_count -= 1
+
+    # If CAP END is sent before the SASL authentication message (_handle_authenticate()),
+    # the server will respond with 904 SASL authentication aborted.
+    # If SASL has been negotiated with the server, Mantaray sends CAP END after the server
+    # has replied with an authentication success/fail message
```
Comment (Akuli) [inline mantaray/received.py:360]: I still have a hard time understanding this comment. How about this? it includes more things that you can easily `git grep`.

```suggestion
    # If "sasl" is in cap_list, Mantaray sends CAP END after the server
    # has replied with RPL_SASLSUCCESS or ERR_SASLFAIL
```

File: mantaray/received.py [hunk 28]
```diff
@@ -316,14 +335,35 @@ def _handle_cap(server_view: views.ServerView, args: list[str]) -> None:
...
             raise ValueError("The server does not support SASL.")
 
+    else:
+        server_view.core.send("CAP END")
+        raise ValueError("Invalid CAP response. Aborting Capability Negotiation.")
+
+    server_view.core.pending_cap_count -= 1
+
+    # If CAP END is sent before the SASL authentication message (_handle_authenticate()),
+    # the server will respond with 904 SASL authentication aborted.
```
Comment (Akuli) [inline mantaray/received.py:358]: These two lines feel to me as a pile of confusing details. How about this, which IMO better expresses what you're really trying to say?

```suggestion
    # If we use SASL, we can't send CAP END until all SASL stuff is done.
```

File: mantaray/views.py [hunk 29]
```diff
@@ -171,6 +180,9 @@ def add_view_selector_tag(self, tag: Literal["new_message", "pinged"]) -> None:
     def destroy_widgets(self) -> None:
         self.textwidget.destroy()
 
+    def get_text(self) -> str:
```
Comment (Akuli) [inline mantaray/views.py:183]:
```suggestion
    # for tests
    def get_text(self) -> str:
```

File: tests/test_commands.py [hunk 30]
```diff
@@ -234,6 +233,65 @@ def test_nickserv_and_memoserv(alice, bob, wait_until):
...
+    alice.on_enter_pressed()
+
+    # Server view
+    wait_until(
+        lambda: "You have been marked as being away\n"
+        in alice.get_server_views()[0].get_text()
+    )
+
+    switch_to_channel_view(alice, "#autojoin")
+
```
Comment (Akuli) [inline tests/test_commands.py:251]: This happens automatically before the test starts

```suggestion
```

File: tests/test_commands.py [hunk 31]
```diff
@@ -234,6 +233,65 @@ def test_nickserv_and_memoserv(alice, bob, wait_until):
...
+    )
+
+
+@pytest.mark.skipif(
+    os.environ["IRC_SERVER"] == "hircd", reason="hircd doesn't support away-notify"
+)
+def test_who_on_join(alice, bob, wait_until):
+    alice.entry.insert(
+        "end", "/part #autojoin"
+    )  # Ensures Alice and Bob are not sharing any channels
```
Comment (Akuli) [inline tests/test_commands.py:276]: Shouldn't the same exact test pass if they are also both on `#autojoin`?

To run the rest of the test in both ways, you can use e.g. `pytest.mark.parametrize("sharing_channels", [True, False])`. Inside the test, you can then put parting `#autojoin` under `if not sharing_channels:`. Mantaray already has a few tests that use `pytest.mark.parametrize` in this way.

**Review by Akuli — COMMENTED** [review]
Thanks for working on this! A few final nits.

File: mantaray/received.py (hunk 25 above)
Comment (ThePhilgrim) [reply mantaray/received.py:346]: @Akuli As the code is working now, acknowledged will ever only contain one capability since we're `REQ`ing one capability at a time. The for loop only adds one capability, but keeping it in a for loop will allow scaling if we later prefer to send several `REQ`s together.

I plan on improving Mantaray's capability negotiation in a new PR shortly.

The only potential issue here (which is currently not an issue) is that if we scale the functionality so that acknowledge can contain more than 1 capability, then `server_view.core.pending_cap_count -= 1` will not suffice. But I think this can be kept for now, since an improvement is coming soon, and atm, this cannot bug.

File: mantaray/received.py (hunk 26 above)
Comment (ThePhilgrim) [reply mantaray/received.py:443]: It should be impossible that it's something else, but I can add it for sanity.

File: mantaray/received.py (hunk 27 above)
Comment (ThePhilgrim) [reply mantaray/received.py:360]: I thought you wanted docstrings/comments to be more high level. That's why I wrote it like this. High level maybe only applies to docstrings?

File: tests/test_commands.py (hunk 31 above)
Comment (ThePhilgrim) [reply tests/test_commands.py:274]: Yes, it should pass. I do the `/part` just to make sure the tag can't in some way be taken from another shared channel. Will add the `parametrize`

File: mantaray/received.py (hunk 25 above)
Comment (Akuli) [reply mantaray/received.py:346]: It would be good if this explanation was a short comment, e.g. `# TODO: splitting currently unnecessary because server sends one capability at a time`

**Review by Akuli — APPROVED** [review]
Thanks! Works great, apart from #238 :)
